CRM: Add React / TypeScript / Jest build configuration

kallehauge opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->
<!-- Would you like this feature to be tested by Beta testers?
Please add testing instructions to projects/plugins/jetpack/to-test.md in a new commit as part of your PR. -->
<!-- a12s: If you have an expected version that you're aiming for the PR to add, please use the Milestone field to communicate it. If you leave it blank, that indicates there isn't a preference. -->

Introduce React / TypeScript / Jest to the Jetpack CRM plugin.
The new configuration is based on how the starter plugin and other Jetpack plugins in the repo structure their code.

I've chosen to add a couple of "not entirely ready for production yet" React components to make the new code manually testable:

1. `AdminPage` which wraps the default Jetpack `AdminPage` component with a CRM specific header and footer.
2. A theoretical example component called `OnboardingWizard` that doesn't do anything else than output hardcoded RNA component.
3. _There's also a `onboarding-wizard-index.jsx` file that only serves the purpose of rendering a new onboarding wizard experience._

Fixes https://github.com/Automattic/zero-bs-crm/issues/2927

## Proposed changes:
<!--- Explain what functional changes your PR includes -->
**Dummy logic**

* Add a new `Automattic\Jetpack_CRM\Onboarding_Wizard\Boostrap` class hidden behind a feature flag to register and render new React components.
  * Add a new onboarind wizard admin page behind a feature flag so we can continuously start creating it (`jetpack_crm_feature_flag_onboarding_wizard_v2`).
* Add `AdminPage` and `OnboardingWizard` React components.

**JS**

* Add new webpack config export that handles our React processing.
* Add `tsconfig.json` so we start to allow type specific imports etc.
* Add Jest configuration and dependencies so we can write tests.
* Re-introduce `eslintrc.js` so we get linting errors when using incorrect text-domain.
* Ignore `/build/` in `.gitignore` since it's the source for our compiled React.

**PHP**

* Add `Automattic\Jetpack\Assets` package so we can register React scripts "the Jetpack way".
* Add `projects/plugins/crm/src/` as a source for the PHP autoloader in `composer.json`, so future internal packages will be autoloaded by the Jetpack Autoloader.

### Other information:

- [x] Have you written new tests for your changes, if applicable?
- [ ] _Have you checked the E2E test CI results, and verified that your changes do not break them?_
- [ ] _Have you tested your changes on WordPress.com, if applicable (if so, you'll see a generated comment below with a script to run)?_

## Jetpack product discussion
<!-- If you're an Automattician, include a shortlink to the p2 discussion with Jetpack Product here. -->
<!-- Make sure any changes to existing products have been discussed and agreed upon -->

There's no conversations about this.

## Does this pull request change what data or activity we track or use?
<!--- If so, please add the "[Status] Needs Privacy Updates" label and explain what changes there are. -->
<!--- Check existing Jetpack support documents for a preview of the information we need. -->

No.

## Testing instructions:
<!-- If you were reviewing this PR, how would you like the instructions to be presented? -->
<!-- Please include detailed testing steps, explaining how to test your change. -->
<!-- Bear in mind that context you working on is not obvious for everyone.  -->
<!-- Adding "simple" configuration steps will help reviewers to get to your PR as quickly as possible. -->
<!-- "Before / After" screenshots can also be very helpful when the change is visual. -->

* Spin up project locally and install latest dependencies
* Run `jetpack build plugins/crm`
* Add `add_filter( 'jetpack_crm_feature_flag_onboarding_wizard_v2', '__return_true' );` to e.g. `0-sandbox.php` so we register a new page that shows our dummy content.
* Go to `/wp-admin/admin.php?page=zerobscrm-obw`
  * Verify that it looks like the screenshot below.
* Run `pnpm run lint-file projects/plugins/crm/` from the root.
  * Verify that no linting issues pops up.
* Navigate to `projects/plugins/crm`
* Run `pnpm run test`
  * Verify that 2 tests successfully pass:
    * `src/js/components/admin-page/test/index.test.jsx`
    * `src/js/components/onboarding-wizard/test/index.test.jsx`

**Screenshot(s)**

![Screenshot 2023-02-21 at 04 01 56](https://user-images.githubusercontent.com/3846700/220237469-ec95a827-3ee4-4ae1-ab48-3e75d8c7b891.jpg)

kallehauge committed:
Add AdminPage react component

And all the dependencies that comes with it.

kallehauge committed:
changelog

kallehauge committed:
Add .eslintrc.js to check for correct text-domain

kallehauge committed:
Correct eslintrc comment

kallehauge committed:
Update composer.lock

This was required because it had become out-of-date from trunk.

kallehauge committed:
Update webpack config with externals i18n workaround

kallehauge commented on projects/plugins/crm/webpack.config.js:
```diff
@@ -9,6 +9,7 @@ const welcomeZBSCSSPattern = '**/css/welcome-to-zbs/*.css';
 
 const alwaysIgnoredFiles = [
 	'**/js/**/*.min.js',
+	'**/src/js/**',
```
We have to ignore `src/js` from our "legacy JS" build to prevent errors due to requiring other dependencies.

kallehauge commented on projects/plugins/crm/webpack.config.js:
```diff
@@ -186,4 +187,58 @@ module.exports = [
+				jetpackWebpackConfig.FileRule(),
+			],
+		},
+		externals: {
+			...jetpackWebpackConfig.externals,
+			jetpackConfig: JSON.stringify( {
+				consumer_slug: 'zero-bs-crm',
+			} ),
+			'@wordpress/i18n': 'global wpI18n',
+			'@wordpress/jp-i18n-loader': 'global jpI18nLoader',
+		},
```
I assume we have to use the same workaround here as we do further up the config?

kallehauge commented on projects/plugins/crm/webpack.config.js:
```diff
@@ -186,4 +187,58 @@ module.exports = [
+		},
+		mode: jetpackWebpackConfig.mode,
+		devtool: jetpackWebpackConfig.isDevelopment ? 'source-map' : false,
+		output: {
+			...jetpackWebpackConfig.output,
+			path: path.resolve( './build' ),
+		},
```
We're going to make a slow transition towards a full React UI which means that we won't have a single index main file which the original intent of this setup (copied from the starter plugin) is meant for.

We can still define different builds in `entry` and the key will be used as the filename, but all of our components would be living in a flat directory structure with the current config. Perhaps we should try and create a directory per build instead?

coder-karen replied:
Yes, we could define different outputs. The Jetpack plugin itself has done similar, an example being in `webpack.config.extensions.js`. An [entry point](https://github.com/Automattic/jetpack/blob/trunk/projects/plugins/jetpack/tools/webpack.config.extensions.js#L175) includes [all blocks in the extensions folder with `view.js` files](https://github.com/Automattic/jetpack/blob/trunk/projects/plugins/jetpack/tools/webpack.config.extensions.js#L52) as I understand it, and the output is [a folder for each block](https://github.com/Automattic/jetpack/blob/trunk/projects/plugins/jetpack/tools/webpack.config.extensions.js#L104) under `_inc/blocks`.

kallehauge replied:
Great suggestion 👏 I've borrowed the concept and implemented it in a simpler way than the Jetpack plugin.
We still look for `view.{js,jsx,ts,tsx}` files amongst our components in a way that makes it compatible with `Automattic\Jetpack\Assets::enqueue_script`.

Example:
`projects/plugins/crm/src/js/components/onboarding-wizard/view.jsx`
=>
`projects/plugins/crm/build/onboarding-wizard/index.asset.php`
`projects/plugins/crm/build/onboarding-wizard/index.css`
`projects/plugins/crm/build/onboarding-wizard/index.css.map`
`projects/plugins/crm/build/onboarding-wizard/index.js`
`projects/plugins/crm/build/onboarding-wizard/index.js.map`
`projects/plugins/crm/build/onboarding-wizard/index.rtl.css`

coder-karen replied:
Nice!

coder-karen reviewed (commented):
It should be possible to make the build look cleaner as you were asking about. That said I've not tested anything at this stage, so just adding initial thoughts (and I'm not familiar with Jest config at this stage).

Adding `...crmWebpackConfig`, at the top of your new export in `webpack.config.js` means you can remove quite a few lines in that file as well:


```
	{
		...crmWebpackConfig,
		entry: {
			'onboarding-wizard': './src/js/onboarding-wizard-index.jsx',
		},
		output: {
			...jetpackWebpackConfig.output,
			path: path.resolve( './build' ),
		},
		plugins: [
			...jetpackWebpackConfig.StandardPlugins( {
				DependencyExtractionPlugin: { injectPolyfill: true },
			} ),
		],
		module: {
			...crmWebpackConfig.module,
			rules: [
				...crmWebpackConfig.module.rules,
				// Handle CSS.
				jetpackWebpackConfig.CssRule( {
					extensions: [ 'css', 'sass', 'scss' ],
					extraLoaders: [ 'sass-loader' ],
				} ),

				// Handle images.
				jetpackWebpackConfig.FileRule(),
			],
		},
	},
```

We could add handling images to the default crmWebpackConfig too.

kallehauge committed:
Merge branch 'trunk' into add/crm/react-ts-jest-jp-autoloader

kallehauge committed:
Update dependencies after trunk merge conflict

kallehauge committed:
Webpack: look for view.{js,jsx,ts,tsx} in React build

kallehauge committed:
Webpack: Reuse CRM configuration in React build

kallehauge commented:
> Adding ...crmWebpackConfig, at the top of your new export in webpack.config.js means you can remove quite a few lines in that file as well

Nice! The only thing that didn't work from your suggestion was the `module` configuration. It kept complaining about because we didn't include the TranspileRule rules:

```
ERROR in ../../js-packages/components/lib/locale/index.ts 13:35
Module parse failed: Unexpected token (13:35)
You may need an appropriate loader to handle this file type, currently no loaders are configured to process this file. See https://webpack.js.org/concepts#loaders
|  * @returns {string} Browser-formatted locale.
|  */
> export const cleanLocale = ( locale: string ) => {
| 	const regex = /^([a-z]{2,3})(_[a-z]{2}|_[a-z][a-z0-9]{4,7})?(?:_.*)?$/i;
|
 @ ../../js-packages/components/index.ts 57:0-58 57:0-58 57:0-58
 @ ./src/js/components/onboarding-wizard/view.jsx 1:0-63 15:53-66
```

> We could add handling images to the default crmWebpackConfig too.

Definitely, but I think that would require us to do proper testing of all existing assets which seems outside the scope of this PR 😄

coder-karen reviewed (commented):
> Nice! The only thing that didn't work from your suggestion was the module configuration. It kept complaining about because we didn't include the TranspileRule rules:

It looks like that is because of the addition of `/lib\//` in the `crmWebpackConfig`'s TranspileRule exclude list. I removed that and the build completed.

I ran throught the testing instructions as well:

> Run jetpack build plugins/jetpack

I'm guessing this should be `jetpack build plugins/crm`?

> Go to /wp-admin/admin.php?page=zerobscrm-obw
> Verify that it looks like the screenshot below.

This didn't work for me - I get an error: 'Sorry, you are not allowed to access this page.'.
I tried adding the filter to `0-sandbox.php` in my local development mu-plugins folder - I also tried adding it via a code snippets plugin. Not too sure how to get this working?

> Run pnpm run lint-file projects/plugins/crm/ from the root.
> Verify that no linting issues pops up.

This worked!

> Navigate to projects/plugins/crm
> Run pnpm run test
> Verify that 2 tests successfully pass:

These two tests pass!

Commit:
Merge branch 'trunk' into add/crm/react-ts-jest-jp-autoloader

kallehauge committed:
Merge branch 'trunk' into add/crm/react-ts-jest-jp-autoloader

kallehauge committed:
Add fresh composer.lock after merge conflicts

kallehauge committed:
Refactor webpack setup pattern to specify build paths

kallehauge committed:
changelog

kallehauge commented on projects/plugins/boost/tsconfig.json:
Super annoying that I'm forced to include this change unless I want to "no verify" that rest of the code after merging trunk into the PR branch, but such is life! 🤷

kallehauge replied:
I have created a new PR to fix this separately: https://github.com/Automattic/jetpack/pull/29496

So please, ignore this for now. It should (hopefully) be fixed soon by merging trunk again 😄

kallehauge commented:
> We could add handling images to the default crmWebpackConfig too.

I think, for this PR, we'll just leave it as-is due to the other changes going on 🙈 

> I'm guessing this should be jetpack build plugins/crm?

Yes, thank you! 😅 

> This didn't work for me - I get an error: 'Sorry, you are not allowed to access this page.'.
I tried adding the filter to 0-sandbox.php in my local development mu-plugins folder - I also tried adding it via a code snippets plugin. Not too sure how to get this working?

Hmm, that's weird. Adding/removing `add_filter( 'jetpack_crm_feature_flag_onboarding_wizard_v2', '__return_true' );` seems to work for me without any additional steps when I go through the instructions.

I'm basically asking you to try and reboot now (sorry) but I just want to make sure 😝 Would you mind trying to run a `wp rewrite flush` to see if that fixes anything? (if you're using JN, then changing the permalink setting should trigger the same logic).

> It looks like that is because of the addition of /lib\// in the crmWebpackConfig's TranspileRule exclude list. I removed that and the build completed.

Yup, you nailed it on the head @coder-karen! IMO we can go one of two directions to handle it:

1. Create separate `jetpackWebpackConfig.TranspileRule` rules because of the way we have setup the pattern wildcards.
2. Be more explicit with our asset patterns so we don't have to use universal global excludes in our generic rulesets.

I've chosen to go with option 2 and be much more explicit in our asset patterns.
IMO it makes the config more clean because we don't have to figure out where e.g. our old "exclude everywhere" patterns are being used (or, maybe, not used anymore), but we instead contain it in the entries functions which should make it easier for us to remove legacy builds in the future as well.

Specifically: because we no longer rely on universal include patterns, we no longer have to add `/lib\//` as an universal exclude path in our generic TranspileRule and can fall back on something generic like the below, which means the original issue no longer exists ( https://github.com/Automattic/jetpack/pull/29062/commits/944ba33f93ab2370500e22606eec608fded623fa ):

```js
jetpackWebpackConfig.TranspileRule( {
	exclude: [ /node_modules\//, /vendor\//, /tests\// ],
} ),
```

This change might seem concerning considering how much trouble we originally had migrating to Webpack, but it's actually pretty easy to test that we build all of the same items.
I've done the following to ensure that we build the exact same files as before and that they're all 100% identical:

1. Removed the rules below from our `.gitignore` which means we can commit all CSS and JS assets in their transpiled version:

```
**/css/*.min.css.map
**/css/**/*.min.css.map
**/css/*.min.rtl.*
**/css/**/*.min.rtl.*
**/css/*.min.css
**/css/welcome-to-zbs/*.min.css
**/css/integrations/*
**/js/*.min.js.map
**/js/**/*.min.js.map
**/js/*.min.js
**/js/welcome-to-zbs/*.min.js
**/js/*.min.js.LICENSE.txt
```

2. Ran `pnpm run clean` from `projects/plugins/crm`.
3. [Used our old build configuration](https://github.com/Automattic/jetpack/blob/a6bda1a8b705aa819023c5a36e5fd7258ab27717/projects/plugins/crm/webpack.config.js).
4. Added all transpiled files to Git.
5. Ran `pnpm run clean` from `projects/plugins/crm` and verified that all files was staged as "deleted" in Git.
6. [Ran a new build using our updated config with explicit rules](https://github.com/Automattic/jetpack/blob/944ba33f93ab2370500e22606eec608fded623fa/projects/plugins/crm/webpack.config.js).
7. Verify that (besides the updated `webpack.config.js` file) we had no other unstaged files in Git which means we have transpiled everything as an exact duplicate of before.

kallehauge committed:
Merge branch 'trunk' into add/crm/react-ts-jest-jp-autoloader

kallehauge committed:
Remove boost changelog

This change was introduced in #29411 instead.

coder-karen reviewed (commented):
> I'm basically asking you to try and reboot now (sorry) but I just want to make sure 

It is working for me now! That's just from running through the steps as they were, but on my new MacBook and freshly installed dev environment, so maybe something was off at my end before.

> Be more explicit with our asset patterns so we don't have to use universal global excludes in our generic rulesets.

That makes sense.

> Removed the rules below from our .gitignore which means we can commit all CSS and JS assets in their transpiled version:

I tried to replicate those steps but I'm not sure I'm running them correctly - when removing lines from .gitignore were you working from this branch (and then following the later steps), or from trunk? If I follow the steps starting from this branch I end up with a lot of new files as well as deleted files at stage 5.

gogdzl commented:
Tested everything and it works as advertised.
I tried to replicate the problems @coder-karen had with .gitignore, but couldn't.
```
diegorodrigues@Diegos-MacBook-Pro jetpack % git status
On branch add/crm/react-ts-jest-jp-autoloader
Your branch is up to date with 'origin/add/crm/react-ts-jest-jp-autoloader'.

nothing to commit, working tree clean
```

kallehauge commented:
> I tried to replicate those steps but I'm not sure I'm running them correctly

I didn't specifically intent for them to be instructions but more of a general idea of how I verified. That being said: let's dig into it @coder-karen !

> when removing lines from .gitignore were you working from this branch (and then following the later steps), or from trunk?

I worked from the this branch and used `git reset --hard origin/add/crm/react-ts-jest-jp-autoloader` when I was done 😄

> If I follow the steps starting from this branch I end up with a lot of new files as well as deleted files at stage 5.

That was also the intended outcome (when we run the `pnpm run clean` script from `projects/plugins/crm` it will delete all of our transpiled files).

The reason why we clean (read: delete) all the files before using the new build configuration is to prevent a false positive result. If we don't delete all the transpiled files when we do a fresh build, then there might be some files that might look like they transpiled successfully, but what in reality they were never touched (this could be because the pattern is missing or something similar).

So if all of the transpiled files (read: files ending in `*.min.css`, etc) were staged as "deleted" after step 5, running the build command in "[step 6. Ran a new build using our updated config with explicit rules](https://github.com/Automattic/jetpack/blob/944ba33f93ab2370500e22606eec608fded623fa/projects/plugins/crm/webpack.config.js)" should generate all of the files again and make your git staging look clean (maybe except for webpack.config.js depending on how you committed it in step 4.)

coder-karen reviewed (approved):
Thanks for clarifying @kallehauge ! I was running 'git add' for step 4, instead of 'git commit', which is why I was ending up with a different result. Just using 'git commit' those steps worked (even though I didn't need to test them :D )
